### DISCUSSION TITLE
Are incremental builds possible?

### DISCUSSION THREAD

**Discussion by rb-determined-ai** [horovod/horovod#2841] [Q&A] [answered]
Are there any tips for how to compile horovod incrementally?  Right now I can run some CMake commands to build the C++ code into importable modules so that e.g. import horovod.torch works fine but import horovod.runner does not, because setup.py is doing some magic stitching of compiled modules with pure python modules.  That magic stitching is eluding me.

I'm trying to work on passing my own unit test for [this fix](https://github.com/horovod/horovod/pull/2724) and on [this issue](https://github.com/horovod/horovod/issues/2806), but I'm a bit baffled so far just trying to get horovod building.

**Comment by rb-determined-ai** [comment, marked as answer]
Got help from @chongxiaoc.  Incremental builds are not supported.  However, I was able to get my build running by removing the `-e` from `pip install -e .` (which is what I was using earlier), and I can speed up the test by adding `-j` to the make flags in the `setup.py` to make it build faster.

**Reply by maxhgerlach** [reply]
Note that with `-j` you can get weird build failures from race conditions when cmake doesn't properly discriminate between the two libraries built in `horovod/common/ops/cuda`.
